_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,281 @@
+---
+layout: post
+title: Of variance and functors
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+Scala's type system allows us to annotate type parameters with their variance: covariant, contravariant, invariant.
+Variance allows us to define the subtyping relationships between type constructors - that is, under which
+conditions `F[A]` is a subtype of `F[B]`.
+
+Similarly in functional programming, there are covariant functors, contravariant functors, and invariant functors. The
+similarity in names is not coincidental.
+
+# Covariance
+The most common example is `List[+A]` which is covariant in its type parameter, denoted by the `+` next to the `A`.
+A type constructor with a covariant type parameter means that if there is a subtyping relationship between the
+type parameter, there is a subtyping relationship between the two instances of the type constructor. This means that
+if we have a `List[Circle]`, we can substitute it anywhere we have a `List[Shape]`.
+
+## Read
+Another example of covariance is in parsing, for example in the following `Read` type class.
+
+```scala
+trait Read[+A] {
+  def read(s: String): Option[A]
+}
+```
+
+It makes sense to make `Read` covariant because if we can read a subtype, then we can read the supertype by
+reading the subtype and throwing away the subtype-specific information. For instance, if we can read a
+`Circle`, we can read a valid `Shape` by reading the `Circle` and ignoring any `Circle`-specific information.
+
+## Array
+A type that cannot safely be made covariant is `Array`. If `Array` were covariant, we could substitute
+an `Array[Circle]` for an `Array[Shape]`. This can get us in a nasty situation.
+
+```scala
+val circles: Array[Circle] = Array.fill(10)(Circle(..))
+val shapes: Array[Shape] = circles // works only if Array is covariant
+shapes(0) = Square(..) // Square is a subtype of Shape
+```
+
+If `Array` was covariant this would compile fine, but fail at runtime. The compiler would accept this
```

Comment:
Could you note here that in Java this compiles, but throws `ArrayStoreException` at runtime?